crates/shared/src/transport/buffered.rs

```diff
@@ -0,0 +1,311 @@
+//! A buffered `Transport` implementation that automatically groups JSON RPC
+//! requests into batches.
+
+use super::MAX_BATCH_SIZE;
+use ethcontract::{
+    jsonrpc::Call,
+    web3::{BatchTransport, Error as Web3Error, RequestId, Transport},
+};
+use futures::{
+    channel::{mpsc, oneshot},
+    future::{self, BoxFuture, FutureExt as _},
+    stream::{self, FusedStream, Stream, StreamExt as _},
+};
+use serde_json::Value;
+use std::{future::Future, sync::Arc, time::Duration};
+use tokio::task::JoinHandle;
+
+/// Buffered transport configuration.
+pub struct Configuration {
+    /// The maximum amount of concurrent batches to send to the node.
+    pub max_concurrent_requests: usize,
```

Comment:
nit: I think `for_each_concurrent` taking an `Into<Option<usize>>` is an unfortunate choice because `Some(0)` (which would mean unlimited parallelism) feels like an invalid state. Making `max_concurrent_requests` a `Into<NonZeroUsize>` to "fix" the API of `for_each_concurrent` seems like too much of a hassle but documenting that passing 0 would result in unlimited parallelism seems reasonable.

Reply:
Good point. I don't like "magic values". I like the idea of an `Option<NonZeroUsize>`.